### PR TITLE
Remove busy waiting in homegrown scheduler when no jobs are spawned

### LINKED ISSUE #33: Busy waiting in homegrown scheduler when no jobs are spawned
Happy new year, all!

Below are the relevant lines of code in the homegrown scheduler.
https://github.com/cmuparlay/parlaylib/blob/6302233a72199b9a0f69e4bdb63928ebf5e8f3cc/include/parlay/scheduler.h#L258-L268

When all deques are empty, `try_steal` will always fail and this loop repeats indefinitely. In each iteration, it will sleep for a short amount of time and wake up again only to fall asleep without doing any computation. The result is high CPU usage despite no actual work being performed.

```c++
#include <iostream>

#include <parlay/scheduler.h>

using namespace std;

int main() {
  parlay::fork_join_scheduler fj;
  int x = 0, y = 0;
  
  cout << "Please enter two integer values: ";
  cin >> x >> y;

  fj.pardo(
    [&]() { x += 1; },
    [&]() { y += 2; }
  );

  cout << x << " " << y << endl;
  return 0;
}
```
Instead of waiting for a fixed amount of time, the scheduler should wait until a deque is (potentially) nonempty.

### PR DESCRIPTION
**Pull request by anivegesana**
This PR uses a simplistic backport of the atomic wait primitive from C++20. It adds an `enabled` variable to the scheduler which determines whether the threads are paused or are able to run jobs. The fork-join scheduler defaults to being paused and turns on whenever `pardo` or `parfor` are called. At the completion of all `pardo` and `parfor` calls, the scheduler is automatically paused to prevent busy waiting in the case that there are no jobs to run.

This code has not yet been tested for efficiency. Help would be appreciated.

Fixes #33

### CODE REVIEW

**Comment by DanielLiamAnderson** [comment]
Thanks for putting this together.  I tried running the benchmarks, but it seems to run into an issue where sometimes (but not consistently), the program fails to terminate. In particular, when I build and run `bench_examples`, the benchmarks all complete, but then the program stalls forever without terminating.  This suggests that the scheduler perhaps is getting stuck during termination. (see image. Program is stuck not terminating at the end)

image

Assuming this can be fixed, there's still some other things that I might be concerned about:

### Performance
- I haven't managed to check the performance yet because of the aforementioned issue, but I suspect that the change may introduce a performance penalty. Specifically, doing an atomic increment/decrement on every pardo could introduce a lot of contention (https://github.com/cmuparlay/parlaylib/pull/34/files#diff-6eda068e7db43c78bd6d29aeab46d53e15f28d2d2f672f422f12e0f0ee5ad0a6R333)
- I'm not sure how efficient condition variables are, but I believe that modern `std::atomic::wait` for C++20 is typically implemented using OS-specific primitives like futexes to make it more efficient than a condition variable.

### Style
- I'm not a fan of inlining the std::atomic child class.  I'd at least put this in a separate file (especially since the type would be replaced when we upgrade to C++20 anyway).

**Comment by anivegesana** [comment]
Hi Daniel. Sorry for the confusion. I marked this PR as a draft because I had a working version that compiles with C++20, but was not able to figure out how to get it to work in C++17. You can test it by changing the 17 in this line to a 20.

https://github.com/anivegesana/parlaylib/blob/master/CMakeLists.txt#L53

I primarily need help in creating a small lightweight backport for the atomic_wait feature. The following repo works, but it isn't exactly lightweight.

https://github.com/ogiroux/atomic_wait

If you guys are planning on updating to C++20 in the future, then maybe we can just wait on this PR. We found a workaround and don't need this feature merged. As for speed, a `PARLAY_INTERACTIVE` macro can be added that controls whether or not the increment/decrement happens. This way, if `std::cin` is not used, then Parlay can just be compiled without this macro and keep the original performance. If `std::cin` is used, then Parlay must compiled with `-DPARLAY_INTERACTIVE` and every `pardo`/`parfor` will check to see if the scheduler is off and turn it on if it isn't on already, as well as turn it off if there are no tasks left.

File: include/parlay/scheduler.h [hunk 1]
```diff
@@ -39,10 +40,20 @@
...
 
+#ifdef PARLAY_INTERACTIVE
+#include "internal/atomic_wait.h"
+#endif
+
 // IWYU pragma: no_include <bits/this_thread_sleep.h>
 
 namespace parlay {
 
+#if __has_cpp_attribute(unlikely)
```
Comment (DanielLiamAnderson) [inline include/parlay/scheduler.h:51]: There's a file, `include/parlay/portability.h`, where these kinds of macros should go.

**Comment by DanielLiamAnderson** [comment]
I think we should actually try to improve the scheduler's performance in this regard. I don't think having a macro that slows the library down is a great solution.  We can probably make it fast and keep the benefits with some more work.

**Comment by shwestrick** [comment]
If it's helpful, both OpenCilk and Rayon appear to be doing something that puts workers to sleep. Presumably they've already gone through the work of verifying that these tricks don't impact the performance of the scheduler too much, so it could be worth trying to adapt one of these for Parlay:
  * https://github.com/OpenCilk/cheetah/blob/release/runtime/worker_sleep.h
  * https://github.com/rayon-rs/rayon/tree/master/rayon-core/src/sleep

**Comment by DanielLiamAnderson** [comment]
Thanks for the pointers!  This is the kind of thing we'd like to implement too.

**Comment by anivegesana** [comment]
Hmm. They seem to do almost the same sort of thing I did. I notice some differences in the memory order of the atomic operations, so it is possible that that is what is causing the inefficiencies. I will look into it.

**Comment by DanielLiamAnderson** [comment]
I think a big difference is that your code is all-or-nothing.  If there's any parallel jobs, then all workers are woken up.  But if there are just a small number of jobs, then ideally only a small number of workers will be awake.

**Comment by DanielLiamAnderson** [comment]
For some concrete data, here is the performance impact of the current change. See attachment. Overall, about 20-35% slower on our main benchmark set.

[PR34.zip](https://github.com/cmuparlay/parlaylib/files/10711377/PR34.zip)

**Comment by anivegesana** [comment]
Hey Daniel,

I updated the code to use a different memory order that seems more reasonable, but I still need to run it on the benchmark set.

**Comment by DanielLiamAnderson** [comment]
I've implemented a similar solution in #52 that solves the same problem but with better performance on the benchmarks.  Thanks for the idea and getting this started!

**Pull request closed without merging**